lib/Internal/Http2Driver.php

```diff
@@ -891,28 +871,35 @@ public function parser(Client $client, $settings = ""): \Generator {
                     $headerArray[$name][] = $value;
                 }
 
-                $ireq = new InternalRequest;
+                $ireq = new Request;
                 $ireq->client = $client;
                 $ireq->streamId = $id;
                 $ireq->trace = $headerList;
                 $ireq->headers = $headerArray;
                 $ireq->protocol = "2.0";
                 $ireq->method = $headerArray[":method"][0];
-                $ireq->uri = $headerArray[":path"][0];
-                $ireq->uriScheme = $headerArray[":scheme"][0] ?? ($client->isEncrypted ? "https" : "http");
+
+                $uri = $headerArray[":path"][0];
+                $scheme = $headerArray[":scheme"][0] ?? ($client->isEncrypted ? "https" : "http");
                 $host = $headerArray[":authority"][0] ?? "";
+
+                if ($host === "") {
+                    $error = self::PROTOCOL_ERROR;
+                    goto connection_error;
+                }
+
                 if (($colon = \strrpos($host, ":")) !== false) {
-                    $ireq->uriPort = (int) \substr($host, $colon + 1);
+                    $port = (int) \substr($host, $colon + 1);
                     $host = \substr($host, 0, $colon);
                 } else {
-                    $ireq->uriPort = $client->serverPort;
+                    $port = $client->serverPort;
                 }
-                $ireq->uriHost = \rawurldecode($host);
-                $uri = $headerArray[":path"][0];
-                if (\strpos($uri, '?') !== false) {
-                    list($uri, $ireq->uriQuery) = \explode("?", $uri, 2);
+
+                if ($host !== null) {
+                    $uri = $scheme . "://" . \rawurldecode($host) . ":" . $port . $uri;
                 }
-                $ireq->uriPath = \rawurldecode($uri);
+
+                $ireq->uri = new Uri($uri);
```

Comment:
Somewhat funny to construct an uri just to have the Uri class deconstruct it again. Possibly we should just have a dedicated static method creating an Uri instance given the actual values?

Reply:
Yes, I think the Uri class could use some improvement in this regard, as the only way to construct one is to have a valid uri string pre-built.

Reply:
Which is totally fine. The `Uri` class would have to validate the parts either way.